llvm/lib/Target/AMDGPU/AMDGPUInsertDelayAlu.cpp

```diff
@@ -362,12 +367,51 @@ class AMDGPUInsertDelayAlu {
             for (MCRegUnit Unit : TRI->regunits(Op.getReg())) {
               auto It = State.find(Unit);
               if (It != State.end()) {
-                Delay.merge(It->second);
-                State.erase(Unit);
+                if (SII->isSALU(MI.getOpcode()) &&
+                    AMDGPU::isSGPR(Op.getReg(), TRI) &&
+                    It->second.VALUCycles > 0) {
+                  deletedCyclesNum = It->second.VALUCycles;
+                  State.erase(Unit);
+                  VALUSALUStall = true;
+                } else {
+                  Delay.merge(It->second);
+                  State.erase(Unit);
+                }
               }
             }
           }
         }
+        unsigned maxCycles = 0;
+        unsigned lastWrite = 0;
+        if (Type != OTHER) {
```

Comment:
Instead of the redundant `Type != OTHER` check, you can move this code into the `if` above.